src/utils.rs

```diff
@@ -0,0 +1,83 @@
+// Compare two iterators to check if the needle is contained in the haystack.
+// Because iteration has to be restarted, the caller must provide two factory
+// functions that can skip to the provided offset.
+pub fn contains_kmp<F1, F2, I: Iterator<Item=char>>(haystack: F1, needle: F2) -> bool
+where F1 : Fn(usize) -> I, F2 : Fn(usize) -> I {
+    let mut left = haystack(0);
+    let mut right = needle(0);
+
+    // Consume the first needle character
+    let first = match right.next() {
+        None => return true,
+        Some(c) => c
+    };
+    
+    let mut next_possible_start = 0;
+    let mut i = 0;
+
+    loop {
+        // Consume left until a char matches the 1st needle char
+        let x = match left.next() {
+            None => return false,
+            Some(c) => c
+        };
+
+        i += 1;
+
+        if x != first {
+            continue;
+        }
+
+        let mut ix = 0;
+
+        loop {
+            let x = match left.next() {
+                None => return false,
+                Some(c) => c
+            };
+            
+            let y = match right.next() {
+                None => return true,
+                Some(c) => c
+            };
+
+            ix += 1;
+
+            if x == first {
+                next_possible_start = ix;
+            }
+
+            if y == x {
+                continue;
+            }
+
+            // Overwrite the two iterators, needle from the top (except the first char)
+            // and haystack from the last seen starting point of a potential match
+            left = haystack(i + next_possible_start - 1);
```

Comment:
I believe the restarting of the iterator and the subsequent skipping will make the worst-case run time at least O(n^2) where n in the length of the haystack. Hard to avoid, though.

Reply:
Also, this doesn't look like an implementation of KMP. Is it?